tree/treeplayer/src/TTreeReader.cxx

```diff
@@ -609,10 +627,34 @@ TTreeReader::EEntryStatus TTreeReader::SetEntryBase(Long64_t entry, bool local)
 
    if ((fEndEntry >= 0 && entry >= fEndEntry) || (fEntry >= fTree->GetEntriesFast())) {
       fEntryStatus = kEntryBeyondEnd;
+      WarnIfFriendsHaveMoreEntries();
       return fEntryStatus;
    }
    fDirector->SetReadEntry(loadResult);
    fEntryStatus = kEntryValid;
+
+   for (auto &&fp : fFriendProxies) {
+      if (fp->GetReadEntry() >= 0)
+         break;
+      // We are going to read an invalid entry from a friend, propagate
+      // this information to the user.
+      const auto *frTree = fp->GetDirector()->GetTree();
+      if (!frTree)
+         break;
+      const std::string frTreeName = dynamic_cast<const TChain *>(frTree)
+                                        ? frTree->GetName()
+                                        : ROOT::Internal::TreeUtils::GetTreeFullPaths(*frTree)[0];
+      // If the friend does not have a TTreeIndex, the cause of a failure reading an entry
+      // is most probably a difference in number of entries between main tree and friend tree
+      if (!fp->IsIndexed()) {
+         std::string msg = "Cannot read entry " + std::to_string(entry) + " from friend tree '" + frTreeName +
+                           "'. The friend tree has less entries than the main tree. Make sure all trees "
+                           "of the dataset have the same number of entries.";
+         throw std::runtime_error{msg};
+      }
+      // TODO: Also handle the opposite situation where the friend has a TTreeIndex
```

Comment:
Potential text:
```
Cannot read entry from the indexed friend tree "friend tree name" corresponding to entry [entry number] in the main tree "[tree name]"
```
And if we want to be pedantic we could even list the index's definition.

Reply:
Thanks, I will take note of this for a follow-up PR where I address the issue with indexed friends. I would keep that strictly separate from this PR as that will also actually require extending the RDataFrame API